Automattic/jetpack
# jetpack plugin: Add support for WordPress 6.7+'s block metadata collection registration

Fixes #39921

## Proposed changes:

Add support for WordPress 6.7+'s block metadata collection registration

WordPress 6.7 introduces `wp_register_block_metadata_collection()`, which improves block registration performance by avoiding individual `block.json` file reads at runtime. This PR:

- Adds a build script to generate a blocks manifest file during the build process
- Registers the manifest on `plugins_loaded` when WordPress 6.7+ is detected
- Adds appropriate tests for manifest generation
- Updates build configuration and gitignore/attributes

This should improve performance for sites running WordPress 6.7+ by reducing filesystem operations during block registration.

See WordPress Core Changeset: https://core.trac.wordpress.org/changeset/59132


### Other information:


## Jetpack product discussion



## Does this pull request change what data or activity we track or use?



## Testing instructions:






* Manually generate a file:
  * `cd projects/plugins/jetpack`
  * `php tools/build-block-manifest.php`
  * Look at `./extensions/blocks/blocks-manifest.php`
* Build process should generate a file
  * Delete `projects/plugins/jetpack/extensions/blocks/blocks-manifest.php`
  * Rebuild the Jetpack plugin `pnpm jetpack build plugins/jetpack`
  * The `blocks-manifest.php` file should now exist
* Unit tests:
  * `cd projects/plugins/jetpack`
  * `./vendor/bin/phpunit tests/php/test_block_manifest.php`
* Complete test:
  * Build and install the jetpack plugin
  * I used: `pnpm jetpack build plugins/jetpack ; pnpm jetpack rsync jetpack /tmp/jetpack ; cd /srv/http/wordpress/wp-content/plugins/ ; cp -R /tmp/jetpack .`
  * Complete installation by connecting to WP.com
  * Inside `wp-includes/blocks.php`, `register_block_type_from_metadata()`, log out if it's going down the `if ( $registry_metadata ) {` path or the `} elseif ( $metadata_file_exists ) {` path. Can log out the `$file_or_folder`. For jetpack blocks, it should be going down the registry_metadata path.
